retrieval/state.go

```diff
@@ -0,0 +1,220 @@
+// This can be used directly with context.WithValue if needed, though the
+// ContextWithState and StateFromContext functions are preferred.
+const ContextKey contextKey = "boxo-retrieval-state"
+
+// MaxFailedProvidersToTrack limits the number of failed provider peer IDs
+// that are kept for diagnostic purposes. This prevents unbounded memory growth
+// while still providing useful debugging information.
+const MaxFailedProvidersToTrack = 3
+
+// RetrievalPhase represents the current phase of content retrieval.
+// Phases progress monotonically - they can only move forward, never backward.
+// This helps identify where in the retrieval process a timeout or failure occurred.
+type RetrievalPhase int
+
+const (
+	// PhaseInitializing indicates the retrieval process has not yet started.
+	PhaseInitializing RetrievalPhase = iota
+	// PhasePathResolution indicates the system is resolving an IPFS path to determine
+	// what content needs to be fetched (e.g., /ipfs/Qm.../path/to/file).
+	PhasePathResolution
+	// PhaseProviderDiscovery indicates the system is finding peers that have the content.
+	PhaseProviderDiscovery
+	// PhaseConnecting indicates the system is establishing connections to providers.
+	PhaseConnecting
+	// PhaseDataRetrieval indicates the system is transferring data to the client.
+	PhaseDataRetrieval
+)
+
+// String returns a human-readable name for the retrieval phase.
+func (p RetrievalPhase) String() string {
+	switch p {
+	case PhaseInitializing:
+		return "initializing"
+	case PhasePathResolution:
+		return "path resolution"
+	case PhaseProviderDiscovery:
+		return "provider discovery"
+	case PhaseConnecting:
+		return "connecting to providers"
+	case PhaseDataRetrieval:
+		return "data retrieval"
+	default:
+		return "unknown"
+	}
+}
+
+// RetrievalState tracks diagnostic information about IPFS content retrieval operations.
+// It is safe for concurrent use and maintains monotonic stage progression.
+// Use ContextWithState to add tracking to a context, and StateFromContext
+// to retrieve the state for updates or inspection
+type RetrievalState struct {
+	// ProvidersFound tracks the number of providers discovered for the content.
+	ProvidersFound atomic.Int32
+	// ProvidersAttempted tracks the number of providers we tried to connect to.
+	ProvidersAttempted atomic.Int32
+	// ProvidersConnected tracks the number of providers successfully connected.
+	ProvidersConnected atomic.Int32
+
+	// phase tracks the current retrieval phase (stored as int32)
+	phase atomic.Int32
+
+	// mu protects failedProviders slice during concurrent access
+	mu sync.RWMutex
+	// Track specific provider failures (limited to first few for brevity)
+	failedProviders []peer.ID
+}
+
+// NewRetrievalState creates a new RetrievalState initialized to PhaseInitializing.
+// The returned state is safe for concurrent use.
+func NewRetrievalState() *RetrievalState {
+	rs := &RetrievalState{
+		failedProviders: make([]peer.ID, 0, MaxFailedProvidersToTrack),
+	}
```

Comment:
I think `failedProviders` would be better left `nil` and append failed providers only when they happen.
```suggestion
	rs := &RetrievalState{}
```